### PR TITLE
Revamped Discovery: Channel based and resilient to errors

### LINKED ISSUE #742: integration: zookeeper random failing in janky
I saw some random failing of the zookeeper test case at least 2 times in `janky'. Not sure if it's caused by non-existed zookeeper image locally on the CI?

Some time this test on `janky` is fine.
Tested newly `run.sh` locally on my machine is fine.

### PR DESCRIPTION
**Pull request by aluzzardi**
- New channel based API
- Resilient to errors
- Cleanup of `Cluster` discovery logic
- Misc cleanup in Discovery API
- Cover more ground in integration tests

ping @abronan @vieux @mavenugo for discovery code
ping @chanwit @snrism @jimmyxian for integration tests

Fixes #742

### CODE REVIEW

**Comment by chanwit** [comment]
@aluzzardi This is fantastic.
The new `file`, `token` and `zk` discovery tests all LGTM.

**Comment by aluzzardi** [comment]
Fixed coverage from `-1.7%` to `+3.77%`. The rules are the rules :)

**Comment by jimmyxian** [comment]
@aluzzardi 
All discovery integration tests are OK! :+1:

**Comment by abronan** [comment]
Reviewed the discovery and the tests, LGTM

ping @vieux

**Comment by snrism** [comment]
@aluzzardi discovery integration tests are fine..the change looks good.

File: cli/cli.go [hunk 1]
```diff
@@ -76,24 +77,37 @@ func Run() {
 			Name:      "list",
 			ShortName: "l",
 			Usage:     "list nodes in a cluster",
+			Flags: []cli.Flag{
```
Comment (vieux) [inline cli/cli.go]: nit: I know it's only used here, but I prefer to have all the flags in `cli/flags.go`

File: cli/cli.go [hunk 2]
```diff
@@ -76,24 +77,37 @@ func Run() {
...
+				cli.StringFlag{
+					Name:  "timeout",
+					Value: "10s",
```
Comment (vieux) [inline cli/cli.go]: nit: can you add a `Usage`

File: cluster/swarm/cluster.go [hunk 3]
```diff
@@ -53,29 +55,22 @@ func NewCluster(scheduler *scheduler.Scheduler, store *state.Store, TLSConfig *t
...
-		}
-
-		entries, err := d.Fetch()
-		if err != nil {
-			log.Fatal(err)
-
+			return nil, err
 		}
-		cluster.newEntries(entries)
+		heartbeat = h
```
Comment (vieux) [inline cluster/swarm/cluster.go]: Here we should check that the heartbeat is >= 1

**Comment by aluzzardi** [comment]
@vieux Done. PTAL

**Comment by vieux** [comment]
LGTM

**Comment by vieux** [comment]
LGTM

**Pull request merged**